crates/nu-cli/src/commands/date/parser.rs

```diff
@@ -0,0 +1,98 @@
+// Modified from chrono::format::scan
+
+use chrono::{FixedOffset, Local};
+
+#[derive(Debug, Clone, PartialEq, Eq, Copy)]
+pub enum ParseErrorKind {
+    /// Given field is out of permitted range.
+    OutOfRange,
+
+    /// The input string has some invalid character sequence for given formatting items.
+    Invalid,
+
+    /// The input string has been prematurely ended.
+    TooShort,
+
+    /// The timezone abbreviation is either invalid or not yet supported.
+    NotSupported,
+}
+
+pub fn get_timezone_offset(s: &str) -> Result<FixedOffset, ParseErrorKind> {
+    let offset_hours = |o| Ok(FixedOffset::east(o * 3600));
+
+    if s.chars().all(|x| x.is_ascii_alphabetic()) {
+        match s.to_lowercase().as_str() {
+            "gmt" | "utc" | "ut" => offset_hours(0),
+            "edt" => offset_hours(-4),
+            "est" | "cdt" => offset_hours(-5),
+            "cst" | "mdt" => offset_hours(-6),
+            "mst" | "pdt" => offset_hours(-7),
+            "pst" => offset_hours(-8),
+            "local" => Ok(*Local::now().offset()),
```

Comment:
Is there a date function or crate we can use to do this? Only allowing utc and US timezones seems limited

Reply:
We could directly use the [parse_from_rfc2822](https://docs.rs/chrono/0.4.19/chrono/struct.DateTime.html#method.parse_from_rfc2822) API in the chrono crate to parse an RFC 2822 timezone. However, this API requires a complete date string input (e.g., "Wed, 18 Feb 2015 23:16:09 GMT") and  therefore it has the overhead of parsing other part that we don't care about (e.g., "Wed, 18 Feb 2015 23:16:09"). Also, the `parse_from_rfc2822` API doesn't support omitting the colon or the minutes part, nor does it support the `UTC` and `LOCAL` timezone, so additional operations need to be added on top of it.

Reply:
Is [parse_from_rfc3339](https://docs.rs/chrono/0.4.19/chrono/struct.DateTime.html#method.parse_from_rfc3339) an option? It seems more intuitive than 2822 to me. Maybe just put another `else if` and check the length of string or the right number of dashes or some minimal validation that it is a 3339 string.

Reply:
@fdncred I do agree that RFC3339 looks more intuitive, but here we are only interested in converting an user-provided timezone to an offset that we can use. For example, with the `date to-timezone PST` or `date to-timezone -08:00` or `date to-timezone -0800` or `date to-timezone -08` command, we would like to know that the expected timezone is UTC-08:00. 

If we use either the [parse_from_rfc2822](https://docs.rs/chrono/0.4.19/chrono/struct.DateTime.html#method.parse_from_rfc2822) or the [parse_from_rfc3339](https://docs.rs/chrono/0.4.19/chrono/struct.DateTime.html#method.parse_from_rfc3339) API for this purpose, we simply prefix a fixed date time string (e.g., "Wed, 18 Feb 2015 23:16:09 " for RFC2822 or "1996-12-19T16:39:57" for RFC3339) to the user-provided timezone string, invoke the API, and then extract the parsed timezone offset. In this sense, the format of the fixed date time string doesn't really matter. Nevertheless, using these APIs incur the cost of parsing the fixed date time string that we don't care about. That's why I extracted the code responsible for parsing the timezone from the `chrono` crate (which includes private functions) and made some modifications to suit our use case (e.g., support `date to-timezone LOCAL`).

The reason I suggested the `parse_from_rfc2822` API is because it supports a few timezone abbreviations while RFC3339 supports none. Based on the discussions in #2324, it seems we should support commands like `date to-timezone UTC` and `date to-timezone PST` so timezone abbreviations (well, some of them) should be supported.

Reply:
Thanks for explaining it to me. I like the idea of using your example, `date to-timezone PST` or `date to-timezone -08:00` or `date to-timezone -0800` or `date to-timezone -08`. It seems like a good compromise to me and it's inclusive of any timezone I think. It appears that your code already works this way, right? If so, I vote we commit this.